File: Mobiles/flutter/lib/features/auth/domain/auth_repository.dart
import 'dart:convert';

import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/api/api_client.dart';
import '../../../core/o11y/errors/o11y_errors.dart';
import '../../../core/o11y/loggers/o11y_logger.dart';
import '../../../core/storage/token_storage.dart';

final authRepositoryProvider = Provider((ref) {
  return AuthRepository(
    apiClient: ref.watch(apiClientProvider),
    tokenStorage: ref.watch(tokenStorageProvider),
    o11yLogger: ref.watch(o11yLoggerProvider),
    o11yErrors: ref.watch(o11yErrorsProvider),
  );
});

class AuthRepository {
  AuthRepository({
    required ApiClient apiClient,
    required TokenStorage tokenStorage,
    required O11yLogger o11yLogger,
    required O11yErrors o11yErrors,
  }) : _apiClient = apiClient,
       _tokenStorage = tokenStorage,
       _o11yLogger = o11yLogger,
       _o11yErrors = o11yErrors;

  final ApiClient _apiClient;
  final TokenStorage _tokenStorage;
  final O11yLogger _o11yLogger;
  final O11yErrors _o11yErrors;

  Future<bool> login(String username, String password) async {
    try {
      final response = await _apiClient.post(
        '/api/users/token/login',
        body: {'username': username, 'password': password},
        endpointName: 'login',
        includeAuth: false,
      );

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        final token = json['token'] as String?;
        if (token != null) {
          // Save to storage - ApiClient will pick up the token via stream subscription
          await _tokenStorage.saveSession(token: token, username: username);
          _o11yLogger.debug(
            'Login successful',
            context: {'username': username},
          );
          return true;
        }
      } else {
        _o11yLogger.warning(
          'Login failed',
          context: {
            'username': username,
            'status_code': response.statusCode.toString(),
          },
        );
      }
      return false;
    } catch (e, stackTrace) {
      _o11yErrors.reportError(
        type: 'API',
        error: 'Failed to login: ${e.toString()}',
        stacktrace: stackTrace,
        context: {'endpoint': 'login', 'username': username},
      );
      return false;
    }
  }

  Future<void> logout() async {
    // Clear from storage - ApiClient will pick up the change via stream subscription
    await _tokenStorage.clearSession();
    _o11yLogger.debug('User logged out');
  }

  /// Loads saved session from persistent storage.
  Future<StoredSession> loadSession() {
    return _tokenStorage.loadSession();
  }
}

File: Mobiles/ios/QuickPizzaIos/Features/Auth/Domain/AuthRepository.swift
import Foundation
import SwiftiePod

let authRepositoryProvider = Provider<AuthServiceProtocol> { pod in
    AuthRepository(
        apiClient: pod.resolve(apiClientProvider),
        tokenStorage: pod.resolve(tokenStorageProvider),
        logger: pod.resolve(loggerProvider),
        tracer: pod.resolve(tracerProvider)
    )
}

/// Concrete implementation of AuthServiceProtocol using the QuickPizza API.
final class AuthRepository: AuthServiceProtocol {
    private let apiClient: APIClientProtocol
    private let tokenStorage: TokenStoring
    private let logger: Logging
    private let tracer: Tracing

    private let authStateContinuation: AsyncStream<Void>.Continuation
    let authStateChanged: AsyncStream<Void>

    init(
        apiClient: APIClientProtocol,
        tokenStorage: TokenStoring,
        logger: Logging,
        tracer: Tracing
    ) {
        self.apiClient = apiClient
        self.tokenStorage = tokenStorage
        self.logger = logger
        self.tracer = tracer

        var continuation: AsyncStream<Void>.Continuation!
        self.authStateChanged = AsyncStream { continuation = $0 }
        self.authStateContinuation = continuation
    }

    var isAuthenticated: Bool {
        tokenStorage.loadSession().isAuthenticated
    }

    var currentUsername: String? {
        tokenStorage.loadSession().username
    }

    func login(username: String, password: String) async -> Bool {
        do {
            return try await tracer.withActiveSpan("auth.login", kind: .client) { span in
                span.setAttribute(key: "user.name", value: username)

                struct LoginBody: Encodable {
                    let username: String
                    let password: String
                }

                let (data, response) = try await apiClient.post(
                    "/api/users/token/login",
                    body: LoginBody(username: username, password: password),
                    includeAuth: false
                )

                span.setAttribute(key: "http.status_code", value: response.statusCode)

                if response.statusCode == 200 {
                    struct LoginResponse: Decodable {
                        let token: String
                    }

                    let loginResponse = try JSONDecoder().decode(LoginResponse.self, from: data)
                    tokenStorage.saveSession(token: loginResponse.token, username: username)
                    apiClient.refreshToken()
                    span.setAttribute(key: "auth.result", value: "success")
                    authStateContinuation.yield(())
                    logger.info("Login successful", attributes: ["username": username])
                    return true
                } else {
                    span.setAttribute(key: "auth.result", value: "failed")
                    logger.warning("Login failed", attributes: [
                        "username": username,
                        "status_code": "\(response.statusCode)",
                    ])
                    return false
                }
            }
        } catch {
            logger.error("Login error", error: error, attributes: ["username": username])
            return false
        }
    }

    func logout() {
        tokenStorage.clearSession()
        apiClient.refreshToken()
        authStateContinuation.yield(())
        logger.info("User logged out")
    }

    func loadSession() -> StoredSession {
        tokenStorage.loadSession()
    }
}
